CRM/Activity/BAO/Activity.php

```diff
@@ -656,246 +656,206 @@ public static function logActivityAction($activity, $logMessage = NULL) {
-      $sql .= "$name $desc,\n";
-      $insertValueSQL[] = $name;
-    }
-
-    // add unique key on activity_id just to be sure
-    // this cannot be primary key because we need that for the auto_increment
-    // fixed_sort_order field
-    $sql .= "
-          UNIQUE KEY ( activity_id )
-        ) ENGINE=HEAP DEFAULT CHARACTER SET utf8 COLLATE utf8_unicode_ci
-        ";
-
-    CRM_Core_DAO::executeQuery($sql);
-
-    $insertSQL = "INSERT INTO {$activityTempTable} (" . implode(',', $insertValueSQL) . " ) ";
-
-    $order = $limit = $groupBy = '';
-    $groupBy = " GROUP BY tbl.activity_id, tbl.activity_type, tbl.case_id, tbl.case_subject ";
-
-    if (!empty($input['sort'])) {
-      if (is_a($input['sort'], 'CRM_Utils_Sort')) {
-        $orderBy = $input['sort']->orderBy();
-        if (!empty($orderBy)) {
-          $order = " ORDER BY $orderBy";
+    if ($params['context'] != 'activity') {
+      $activityParams['status_id'] = CRM_Core_PseudoConstant::getKey(__CLASS__, 'status_id', 'Scheduled');
+    }
+
+    // activity type ID clause
+    if (!empty($params['activity_type_id'])) {
+      if (is_array($params['activity_type_id'])) {
+        foreach ($params['activity_type_id'] as $idx => $value) {
+          $params['activity_type_id'][$idx] = CRM_Utils_Type::escape($value, 'Positive');
         }
+        $activityParams['activity_type_id'] = array('IN' => $params['activity_type_id']);
       }
-      elseif (trim($input['sort'])) {
-        $sort = CRM_Utils_Type::escape($input['sort'], 'String');
-        $order = " ORDER BY $sort ";
+      else {
+        $activityParams['activity_type_id'] = CRM_Utils_Type::escape($params['activity_type_id'], 'Positive');
       }
     }
+    elseif (!empty($activityTypes) && count($activityTypes)) {
+      $activityParams['activity_type_id'] = array('IN' => array_keys($activityTypes));
+    }
 
-    if (empty($order)) {
-      // context = 'activity' in Activities tab.
-      $order = (CRM_Utils_Array::value('context', $input) == 'activity') ? " ORDER BY tbl.activity_date_time desc " : " ORDER BY tbl.status_id asc, tbl.activity_date_time asc ";
+    if (!empty($params['activity_type_exclude_id'])) {
+      if (is_array($params['activity_type_exclude_id'])) {
+        foreach ($params['activity_type_exclude_id'] as $idx => $value) {
+          $params['activity_type_exclude_id'][$idx] = CRM_Utils_Type::escape($value, 'Positive');
+        }
+        $activityParams['activity_type_id'] = array('NOT IN' => $params['activity_type_exclude_id']);
+      }
+      else {
+        $activityParams['activity_type_id'] = array('!=' => CRM_Utils_Type::escape($params['activity_type_exclude_id'], 'Positive'));
+      }
     }
 
-    if (!empty($input['rowCount']) &&
-      $input['rowCount'] > 0
+    if (!empty($params['rowCount']) &&
+      $params['rowCount'] > 0
     ) {
-      $limit = " LIMIT {$input['offset']}, {$input['rowCount']} ";
+      $activityParams['options']['limit'] = $params['rowCount'];
     }
```

Comment:
I think we should include `limit = 0` when we intend to fetch only count($getCount = 0). Similar to -

    if ($getCount) {
      $activityParams['options']['limit'] = 0;
    }

As I was able to retrieve a max of 25 count on a summary page of a contact having above 100 activities.

<img width="486" alt="image" src="https://cloud.githubusercontent.com/assets/5929648/24692506/020b4da6-19f7-11e7-9500-8cf2af81d506.png">

After patch - 

<img width="475" alt="image" src="https://cloud.githubusercontent.com/assets/5929648/24692525/1dc45be6-19f7-11e7-9e07-217cbe77811c.png">